write-fonts/src/tables/cmap.rs

```diff
@@ -172,9 +201,27 @@ impl Cmap {
     /// subtables, respectively for the Basic Multilingual Plane and Full Unicode Repertoire.
     ///
     /// Also see: <https://learn.microsoft.com/en-us/typography/opentype/spec/recom#cmap-table>
-    pub fn from_mappings(mappings: impl IntoIterator<Item = (char, GlyphId)>) -> Cmap {
+    pub fn from_mappings(
+        mappings: impl IntoIterator<Item = (char, GlyphId)>,
+    ) -> Result<Cmap, MappingError> {
         let mut mappings: Vec<_> = mappings.into_iter().collect();
-        mappings.sort();
+
+        let mut error = None;
+        mappings.sort_by(|(a, gid1), (b, gid2)| {
+            let ord = a.cmp(b);
+            if ord == std::cmp::Ordering::Equal && gid1 != gid2 {
+                let (gid1, gid2) = if gid1 > gid2 {
+                    (*gid2, *gid1)
+                } else {
+                    (*gid1, *gid2)
+                };
+                error = Some(MappingError::ConflictingMappings { ch: *a, gid1, gid2 })
+            }
+            ord
+        });
+        if let Some(err) = error {
+            return Err(err);
+        }
```

Comment:
this works (and is nice in that we only need a single pass) but it's also a bit tricky to follow, and this code is not particularly hot. I'd personally write this in two passes, like (and now I'm mainly just entertaining myself)

```suggestion
        mappings.sort();
        if let Some((ch, gid1, gid2)) =
            mappings
                .iter()
                .zip(mappings.iter().skip(1))
                .find_map(|((c1, g1), (c2, g2))| {
                    (c1 == c2 && g1 != g2).then(|| (*c1, *g1.min(g2), *g1.max(g2)))
                })
        {
            return Err(MappingError::ConflictingMappings { ch, gid1, gid2 });
        }
```

I'm also wondering how we handle the situation where the mapping just includes duplicates? I think that we should actually be calling `mappings.dedup()` after we sort...

Reply:
Yeah, easy to dedup after sorting, assuming this code isn't hot enough to mind doing extra passes.